### ISSUE TITLE
Incorrect redundant bracket warning with Template Haskell and type applications

### ISSUE THREAD

**Issue by expipiplus1** [ndmitchell/hlint#1153]
Hlint says that this bracket is redundant, but without it the parse returns the operator `@$` instead, and I believe TypeApplications requires no space between the `@` and the type, hence the bracket is necessary here:

```haskell
foo = [|get @($r)|]
```

Almost the opposite of https://github.com/ndmitchell/hlint/issues/558

**Comment by ndmitchell** [comment]
Indeed, a curious find. We have the blanket rule that if an atom is surrounded by brackets then those brackets are redundant. But that's clearly not the case here, so I think we'll need to figure out a better way to express this concept.

**Comment by ndmitchell** [comment]
After taking a closer look, its a weird one, because the `@` is an expression, but the `($r)` is a type. I've made it so that splice types aren't necessarily atomic, which fixes this bug. It might remove some potential warnings, but given the number of splices in types its probably not that high.

**Comment by expipiplus1** [comment]
Thank you!
